File: src/main/java/com/freya02/botcommands/internal/CommandsBuilderImpl.java
package com.freya02.botcommands.internal;

import com.freya02.botcommands.api.BContext;
import com.freya02.botcommands.api.DefaultMessages;
import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.api.RegistrationListener;
import com.freya02.botcommands.api.annotations.JDAEventListener;
import com.freya02.botcommands.api.application.ApplicationCommand;
import com.freya02.botcommands.api.application.CommandPath;
import com.freya02.botcommands.api.application.context.annotations.JDAMessageCommand;
import com.freya02.botcommands.api.application.context.annotations.JDAUserCommand;
import com.freya02.botcommands.api.application.slash.annotations.AutocompletionHandler;
import com.freya02.botcommands.api.application.slash.annotations.JDASlashCommand;
import com.freya02.botcommands.api.prefixed.TextCommand;
import com.freya02.botcommands.api.prefixed.annotations.JDATextCommand;
import com.freya02.botcommands.api.waiter.EventWaiter;
import com.freya02.botcommands.internal.application.ApplicationCommandListener;
import com.freya02.botcommands.internal.application.ApplicationCommandsBuilder;
import com.freya02.botcommands.internal.application.ApplicationUpdaterListener;
import com.freya02.botcommands.internal.application.slash.autocomplete.AutocompletionHandlersBuilder;
import com.freya02.botcommands.internal.components.ComponentsBuilder;
import com.freya02.botcommands.internal.events.EventListenersBuilder;
import com.freya02.botcommands.internal.prefixed.CommandListener;
import com.freya02.botcommands.internal.prefixed.HelpCommand;
import com.freya02.botcommands.internal.prefixed.PrefixedCommandsBuilder;
import com.freya02.botcommands.internal.prefixed.TextCommandInfo;
import com.freya02.botcommands.internal.utils.ClassInstancer;
import com.freya02.botcommands.internal.utils.ReflectionUtils;
import com.freya02.botcommands.internal.utils.Utils;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.sharding.ShardManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

public final class CommandsBuilderImpl {
	private static final Logger LOGGER = Logging.getLogger();
	private static final List<Class<? extends Annotation>> applicationMethodAnnotations = List.of(JDASlashCommand.class, JDAMessageCommand.class, JDAUserCommand.class);

	private final PrefixedCommandsBuilder prefixedCommandsBuilder;
	private final ApplicationCommandsBuilder applicationCommandsBuilder;
	private final EventListenersBuilder eventListenersBuilder;
	private final AutocompletionHandlersBuilder autocompletionHandlersBuilder;

	private final ComponentsBuilder componentsBuilder;

	private final BContextImpl context;
	private final Set<Class<?>> classes;

	private final boolean usePing;
	
	private final List<Class<?>> ignoredClasses = new ArrayList<>();

	public CommandsBuilderImpl(BContextImpl context, List<Long> slashGuildIds, Set<Class<?>> classes) {
		if (classes.isEmpty())
			LOGGER.warn("No classes have been found, make sure you have at least one search path");

		this.context = context;
		this.prefixedCommandsBuilder = new PrefixedCommandsBuilder(context);
		this.componentsBuilder = new ComponentsBuilder(context);
		
		this.usePing = context.getPrefixes().isEmpty();
		if (usePing) LOGGER.info("No prefix has been set, using bot ping as prefix");

		this.classes = classes;
		this.applicationCommandsBuilder = new ApplicationCommandsBuilder(context, slashGuildIds);

		this.eventListenersBuilder = new EventListenersBuilder(context);
		this.autocompletionHandlersBuilder = new AutocompletionHandlersBuilder(context);
	}

	private void buildClasses() throws Exception {
		classes.removeIf(c -> {
			try {
				return !ReflectionUtils.isInstantiable(c);
			} catch (IllegalAccessException | InvocationTargetException e) {
				LOGGER.error("An error occurred while trying to find if a class is instantiable", e);

				throw new RuntimeException("An error occurred while trying to find if a class is instantiable", e);
			}
		});

		for (Class<?> aClass : classes) {
			processClass(aClass);
		}

		if (!context.isHelpDisabled()) {
			processClass(HelpCommand.class);

			final TextCommandInfo helpInfo = context.findFirstCommand(CommandPath.of("help"));
			if (helpInfo == null) throw new IllegalStateException("HelpCommand did not build properly");

			final HelpCommand help = (HelpCommand) helpInfo.getInstance();
			help.generate();
		}

		prefixedCommandsBuilder.postProcess();

		if (context.getComponentManager() != null) {
			//Load button listeners
			for (Class<?> aClass : classes) {
				componentsBuilder.processClass(aClass);
			}
		} else {
			LOGGER.info("ComponentManager is not set, the Components API, paginators and menus won't be usable");
		}

		applicationCommandsBuilder.postProcess();

		if (context.getComponentManager() != null) {
			componentsBuilder.postProcess();
		}

		eventListenersBuilder.postProcess();

		autocompletionHandlersBuilder.postProcess();

		context.getRegistrationListeners().forEach(RegistrationListener::onBuildComplete);

		LOGGER.info("Finished registering all commands");
	}

	private void processClass(Class<?> aClass) throws InvocationTargetException, IllegalAccessException, InstantiationException {
		if (!Modifier.isAbstract(aClass.getModifiers()) && !Modifier.isInterface(aClass.getModifiers())) {
			boolean foundSomething = false;

			//Search for methods annotated with a compatible annotation
			for (Method method : aClass.getDeclaredMethods()) {
				foundSomething |= processMethod(method);
			}

			if (!foundSomething) {
				ignoredClasses.add(aClass);
			}
		}
	}

	@SuppressWarnings("unchecked")
	@Nullable
	private <T> T tryInstantiateMethod(@NotNull Class<? extends Annotation> requiredAnnotation,
	                                   @NotNull Class<T> requiredClass,
	                                   @NotNull String requiredClassDesc,
	                                   @NotNull Method method) throws InvocationTargetException, InstantiationException, IllegalAccessException {
		final Class<?> declaringClass = method.getDeclaringClass();

		if (method.isAnnotationPresent(requiredAnnotation)) {
			if (!requiredClass.isAssignableFrom(declaringClass))
				throw new IllegalArgumentException("Method " + Utils.formatMethodShort(method) + " is annotated with @" + requiredAnnotation.getSimpleName() + " but its class does not extend " + requiredClass.getSimpleName());

			final T annotatedInstance = (T) ClassInstancer.instantiate(context, declaringClass);

			if (!method.canAccess(annotatedInstance))
				throw new IllegalStateException(requiredClassDesc + " " + Utils.formatMethodShort(method) + " is not public");

			if (Modifier.isStatic(method.getModifiers()))
				throw new IllegalStateException(requiredClassDesc + " " + Utils.formatMethodShort(method) + " is static");

			return annotatedInstance;
		}

		return null;
	}

	private boolean processMethod(Method method) throws InvocationTargetException, InstantiationException, IllegalAccessException {
		for (Class<? extends Annotation> annotation : applicationMethodAnnotations) {
			final ApplicationCommand applicationCommand = tryInstantiateMethod(annotation, ApplicationCommand.class, "Application command", method);

			if (applicationCommand != null) {
				applicationCommandsBuilder.processApplicationCommand(applicationCommand, method);

				return true;
			}
		}

		final TextCommand textCommand = tryInstantiateMethod(JDATextCommand.class, TextCommand.class, "Text command", method);
		if (textCommand != null) {
			prefixedCommandsBuilder.processPrefixedCommand(textCommand, method);

			return true;
		}

		final Object eventListener = tryInstantiateMethod(JDAEventListener.class, Object.class, "JDA event listener", method);
		if (eventListener != null) {
			eventListenersBuilder.processEventListener(eventListener, method);

			return true;
		}

		final Object autocompletionHandler = tryInstantiateMethod(AutocompletionHandler.class, Object.class, "Slash command auto completion", method);
		if (autocompletionHandler != null) {
			autocompletionHandlersBuilder.processHandler(autocompletionHandler, method);

			return true;
		}

		return false;
	}

	/**
	 * Builds the command listener and automatically registers all listener to the JDA instance
	 *
	 * @param jda The JDA instance of your bot
	 */
	public void build(JDA jda) throws Exception {
		if (jda.getShardInfo().getShardId() != 0) {
			LOGGER.warn("A shard other than 0 was passed to CommandsBuilder#build, shard 0 is needed to handle DMing exceptions, manually retrieving shard 0...");

			final ShardManager manager = jda.getShardManager();
			if (manager == null) throw new IllegalArgumentException("Unable to retrieve Shard 0 as shard manager is null");

			jda = manager.getShardById(0);

			if (jda == null) throw new IllegalArgumentException("Unable to retrieve Shard 0");
		}

		if (jda.getStatus() != JDA.Status.CONNECTED) {
			try {
				LOGGER.warn("JDA should already be ready when you call #build on CommandsBuilder !");

				jda.awaitReady();
			} catch (InterruptedException e) {
				throw new RuntimeException("CommandsBuilder got interrupted while waiting for JDA to be ready", e);
			}
		}

		final List<GatewayIntent> intents = List.of(
				GatewayIntent.GUILD_MESSAGES
		);
		if (!jda.getGatewayIntents().containsAll(intents)) {
			throw new IllegalStateException("JDA must have these intents enabled: " + intents.stream().map(Enum::name).collect(Collectors.joining(", ")));
		}

		setupContext(jda);

		ReflectionUtils.scanAnnotations(classes);

		buildClasses();

		context.addEventListeners(
				new EventWaiter(jda),
				new CommandListener(context),
				new ApplicationUpdaterListener(context),
				new ApplicationCommandListener(context)
		);

		if (!ignoredClasses.isEmpty()) {
			LOGGER.trace("Ignored classes in search paths:");
			for (Class<?> ignoredClass : ignoredClasses) {
				LOGGER.trace("\t{}", ignoredClass.getName());
			}
		}

		ConflictDetector.detectConflicts();
	}

	private void setupContext(JDA jda) {
		context.setJDA(jda);
		if (usePing) {
			context.addPrefix("<@" + jda.getSelfUser().getId() + "> ");
			context.addPrefix("<@!" + jda.getSelfUser().getId() + "> ");
		}

		context.registerConstructorParameter(BContext.class, ignored -> context);
		context.registerCommandDependency(BContext.class, () -> context);
		context.registerCustomResolver(BContext.class, ignored -> context);

		context.registerConstructorParameter(JDA.class, ignored -> jda);
		context.registerCommandDependency(JDA.class, () -> jda);
		context.registerCustomResolver(JDA.class, ignored -> jda);

		context.setDefaultMessageProvider(new Function<>() {
			private final Map<Locale, DefaultMessages> localeDefaultMessagesMap = new HashMap<>();

			@Override
			public DefaultMessages apply(Guild guild) {
				final Locale effectiveLocale = context.getEffectiveLocale(guild);

				return localeDefaultMessagesMap.computeIfAbsent(effectiveLocale, DefaultMessages::new);
			}
		});
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/CommandsBuilderImpl.kt
package com.freya02.botcommands.internal

import com.freya02.botcommands.annotations.api.application.context.annotations.JDAMessageCommand
import com.freya02.botcommands.annotations.api.application.context.annotations.JDAUserCommand
import com.freya02.botcommands.annotations.api.application.slash.annotations.JDASlashCommand
import com.freya02.botcommands.api.BContext
import com.freya02.botcommands.api.Logging
import com.freya02.botcommands.api.annotations.Declaration
import com.freya02.botcommands.api.application.CommandPath
import com.freya02.botcommands.internal.application.ApplicationCommandListener
import com.freya02.botcommands.internal.application.ApplicationCommandsBuilder
import com.freya02.botcommands.internal.application.ApplicationUpdaterListener
import com.freya02.botcommands.internal.application.slash.autocomplete.AutocompletionHandlersBuilder
import com.freya02.botcommands.internal.components.ComponentsBuilder
import com.freya02.botcommands.internal.modals.ModalHandlersBuilder
import com.freya02.botcommands.internal.prefixed.CommandListener
import com.freya02.botcommands.internal.prefixed.HelpCommand
import com.freya02.botcommands.internal.prefixed.PrefixedCommandsBuilder
import com.freya02.botcommands.internal.utils.ClassInstancer
import dev.minn.jda.ktx.events.CoroutineEventManager
import dev.minn.jda.ktx.events.getDefaultScope
import kotlinx.coroutines.cancel
import net.dv8tion.jda.api.events.ShutdownEvent
import java.lang.reflect.InvocationTargetException
import java.util.*
import kotlin.reflect.KClass
import kotlin.reflect.KFunction
import kotlin.reflect.full.*
import kotlin.reflect.jvm.javaMethod
import kotlin.reflect.jvm.jvmErasure
import kotlin.time.Duration.Companion.minutes

private val LOGGER = Logging.getLogger()

@Deprecated("To be removed")
class CommandsBuilderImpl(context: BContextImpl, packages: Set<String>, userClasses: Set<Class<*>>, slashGuildIds: List<Long>) {
    private val prefixedCommandsBuilder: PrefixedCommandsBuilder

    //	private final EventListenersBuilder eventListenersBuilder; //TODO event listener
    private val applicationCommandsBuilder: ApplicationCommandsBuilder
    private val autocompletionHandlersBuilder: AutocompletionHandlersBuilder
    private val modalHandlersBuilder: ModalHandlersBuilder
    private val componentsBuilder: ComponentsBuilder
    private val context: BContextImpl
    private val classes: List<KClass<*>>
    private val usePing: Boolean
    private val ignoredClasses: MutableList<KClass<*>> = ArrayList()

    init {
        this.classes = findClasses(packages, userClasses)
        if (classes.isEmpty()) LOGGER.warn("No classes have been found, make sure you have at least one search path")
        this.context = context
        this.prefixedCommandsBuilder = PrefixedCommandsBuilder(context)
        this.componentsBuilder = ComponentsBuilder(context)
        this.usePing = context.prefixes.isEmpty()
        if (usePing) LOGGER.info("No prefix has been set, using bot ping as prefix")

        this.applicationCommandsBuilder = ApplicationCommandsBuilder(context, slashGuildIds)

//		this.eventListenersBuilder = new EventListenersBuilder(context); //TODO event listener
        this.autocompletionHandlersBuilder = AutocompletionHandlersBuilder(context)
        this.modalHandlersBuilder = ModalHandlersBuilder(context)
    }

    private fun findClasses(packages: Set<String>, userClasses: Set<Class<*>>): List<KClass<*>> {
//        val scanResult = ReflectionMetadata.runScan(packages)
//
//        val classes = scanResult
//            .allClasses
//            .filter(ReflectionUtilsKt::isInstantiable)
//            .loadClasses()
//            .map(Class<*>::kotlin) + userClasses.map(Class<*>::kotlin)
//
//        ReflectionMetadata.readAnnotations(scanResult)

        return listOf()
    }

    @Throws(Exception::class)
    private fun buildClasses() {
        for (aClass in classes) {
            processClass(aClass)
        }

        if (!context.isHelpDisabled) {
            check(context.findFirstCommand(CommandPath.of("help")) == null) { "Help command was detected before build, and default help is not disabled, consider disabling it with TextCommandsBuilder#disableHelpCommand" }

            processClass(HelpCommand::class)

            val helpInfo = context.findFirstCommand(CommandPath.of("help"))
                ?: throw IllegalStateException("HelpCommand did not build properly")

            val help = helpInfo.instance as HelpCommand
            help.generate()
        }

        prefixedCommandsBuilder.postProcess()

//        if (context.componentManager != null) { //TODO components
//            //Load button listeners
//            for (aClass in classes) {
//                componentsBuilder.processClass(aClass)
//            }
//        } else {
//            LOGGER.info("ComponentManager is not set, the Components API, paginators and menus won't be usable")
//        }

        applicationCommandsBuilder.postProcess()

        if (context.componentManager != null) {
            componentsBuilder.postProcess()
        }

//		eventListenersBuilder.postProcess(); //TODO event listener
//        autocompletionHandlersBuilder.postProcess() //TODO autocomplete
//        modalHandlersBuilder.postProcess() //TODO modals

        context.registrationListeners.forEach { it.onBuildComplete() }

        LOGGER.info("Finished registering all commands")
    }

    @Throws(InvocationTargetException::class, IllegalAccessException::class, InstantiationException::class)
    private fun processClass(aClass: KClass<*>) {
        if (!aClass.isAbstract && !aClass.java.isInterface) {
            var foundSomething = false

            //Search for methods annotated with a compatible annotation
            for (method in aClass.memberFunctions) {
                foundSomething = foundSomething or processMethod(aClass, method)
            }

            if (!foundSomething) {
                ignoredClasses.add(aClass)
            }
        }
    }

    @OptIn(ExperimentalStdlibApi::class)
    @Throws(InvocationTargetException::class, InstantiationException::class, IllegalAccessException::class)
    private inline fun <reified T : Any> tryInstantiateMethod(
        requiredAnnotation: KClass<out Annotation>,
        requiredClassDesc: String,
        method: KFunction<*>
    ): T? {
        val declaringClass = method.javaMethod!!.declaringClass.kotlin

        if (method.findAnnotations(requiredAnnotation).isNotEmpty()) {
            requireUser(T::class.isSuperclassOf(declaringClass), method) {
                "Method is annotated with @${requiredAnnotation.java.simpleName} but its class does not extend ${T::class.simpleName}"
            }

            val annotatedInstance = ClassInstancer.instantiate(context, declaringClass) as T
            requireUser(method.isPublic, method) {
                "$requiredClassDesc is not public"
            }

            requireUser(!method.isStatic, method) {
                "$requiredClassDesc is static"
            }

            return annotatedInstance
        }

        return null
    }

    @Throws(InvocationTargetException::class, InstantiationException::class, IllegalAccessException::class)
    private fun processMethod(clazz: KClass<*>, function: KFunction<*>): Boolean { //TODO pass to method above
//        for (annotation in applicationMethodAnnotations) {
//            val applicationCommand =
//                tryInstantiateMethod<ApplicationCommand>(annotation, "Application command", method)
//            if (applicationCommand != null) {
//                applicationCommandsBuilder.processApplicationCommand(applicationCommand, method)
//                return true
//            }
//        }
//
//        val textCommand = tryInstantiateMethod<TextCommand>(JDATextCommand::class, "Text command", method)
//        if (textCommand != null) {
//            prefixedCommandsBuilder.processPrefixedCommand(textCommand, method)
//            return true
//        }
//
//        val eventListener = tryInstantiateMethod<Any>(
//            JDAEventListener::class,
//            "JDA event listener",
//            method
//        ) //TODO event listener
//        if (eventListener != null) {
//            eventListenersBuilder.processEventListener(eventListener, method)
//            return true
//        }
//
//        val autocompletionHandler = tryInstantiateMethod<Any>(
//            AutocompletionHandler::class,
//            "Slash command auto completion",
//            method
//        )
//        if (autocompletionHandler != null) {
//            autocompletionHandlersBuilder.processHandler(autocompletionHandler, method)
//            return true
//        }
//
//        val modalHandler = tryInstantiateMethod<Any>(ModalHandler::class, "Modal handler", method)
//        if (modalHandler != null) {
//            modalHandlersBuilder.processHandler(modalHandler, method)
//            return true
//        }

        if (function.hasAnnotation<Declaration>()) {
            val args = function.valueParameters.map {
                val value = context.getMethodParameterSupplier(it.type.jvmErasure.java).supply()
                requireUser(value != null, function) {
                    "Requested a parameter '${it.bestName}' with type '${it.type}' but no method parameter supplier was registered for it"
                }

                value
            }.toTypedArray()

            function.call(ClassInstancer.instantiate(context, clazz), *args)

            return true
        }

        return false
    }

    /**
     * Builds the command listener and automatically registers all listener to the JDA instance
     *
     * @param jda The JDA instance of your bot
     */
    @Throws(Exception::class)
    fun build(manager_: CoroutineEventManager?) {
        val manager: CoroutineEventManager = manager_ ?: run {
            val scope = getDefaultScope()
            CoroutineEventManager(scope, 1.minutes).apply {
                listener<ShutdownEvent> {
                    scope.cancel()
                }
            }
        }

        setupContext()

        buildClasses()

        context.addEventListeners( //TODO remove once everything is registered via KTX extensions
            CommandListener(context, usePing),
            ApplicationUpdaterListener(context),
            ApplicationCommandListener(context)
        )

        if (ignoredClasses.isNotEmpty()) {
            LOGGER.trace("Ignored classes in search paths:")
            for (ignoredClass in ignoredClasses) {
                LOGGER.trace("\t{}", ignoredClass.simpleName)
            }
        }

        ConflictDetector.detectConflicts()
    }

    private fun setupContext() {
        context.registerConstructorParameter(BContext::class.java) { context } //TODO maybe unify this stuff ?
        context.registerCommandDependency(BContext::class.java) { context }
        context.registerCustomResolver(BContext::class.java) { _, _, _ -> context } //TODO merge ?
        context.registerMethodParameterSupplier(BContext::class.java) { context }

        context.setDefaultMessageProvider(DefaultMessagesFunction())
    }

    companion object {
        private val applicationMethodAnnotations = listOf(JDASlashCommand::class, JDAMessageCommand::class, JDAUserCommand::class)
    }
}